### PR TITLE
[Feature:Autograding] Add Prolog image

### PR DESCRIPTION
**Pull request by ION606**
### Please check if the PR fulfills these requirements:

* [x] Tests for the changes have been added/updated (if possible)
* [x] Documentation has been updated/added if relevant

### What is the current behavior?

none

### What is the new behavior?
we can now run prolog files

### Other information?

this is not a breaking change and I tested it using:
```sh
docker build -t prolog-environment .
docker run -it prolog-environment
  > swipl
      > ?- write('Hello, Prolog!'), nl.
      > ?- halt.
```

### CODE REVIEW

**Comment by ION606** [comment]
I'm still a bit confused on how to contribute to this repository, so if this is wrong please let me know.

File: dockerfiles/prolog/latest/Dockerfile
Comment (cjreed121) [inline dockerfiles/prolog/latest/Dockerfile:1]: We prefer our docker images to base off of Ubuntu. Should be able to easily swap it for `ubuntu:22.04`.

File: dockerfiles/prolog/latest/Dockerfile
Comment (cjreed121) [inline dockerfiles/prolog/latest/Dockerfile:16]: Can these just be combined? Would make build time faster.

**Review by cjreed121 — CHANGES_REQUESTED** [review]
It's typically preferred to try and lock dependencies for good reproducibility in builds. Can we lock swi-prolog to it's current major version? And if so, it would probably be good to have the tag name be that major version rather than latest.

**Comment by ION606** [comment]
> It's typically preferred to try and lock dependencies for good reproducibility in builds. Can we lock swi-prolog to it's current major version? And if so, it would probably be good to have the tag name be that major version rather than latest.

I'm on mobile at the moment but I'll fix the version once I can test it on desktop

**Comment by ION606** [comment]
> It's typically preferred to try and lock dependencies for good reproducibility in builds. Can we lock swi-prolog to it's current major version? And if so, it would probably be good to have the tag name be that major version rather than latest.

swipl seems to not allow installing versions? I think? I can install the dpkg manually (like they do [here](https://github.com/SWI-Prolog/docker-swipl/blob/4c15bb03b82ab39fb3bdf6579ad66014111faa11/9.2.6/bookworm/Dockerfile)) but I don't know if that would take too long to get and compile and run

File: dockerfiles/prolog/latest/Dockerfile
Comment (ION606) [reply dockerfiles/prolog/latest/Dockerfile:1]: oh I just copied what the verilog file does but yes I'll swap it

**Comment by cjreed121** [comment]
Which version of prolog are you looking to install on this Docker image? Ubuntu repos for 22.04 only have up to 8.4.2. If you wanted to specify this version for apt you would use `swi-prolog=8.4.2+dfsg-2ubuntu1` which was retrieved from https://packages.ubuntu.com/jammy/swi-prolog.

If you wanted a later version, you could look into using the PPA provided by swi-prolog. dpkg is definitely an option but let's try not to complicate this image too much.

File: dockerfiles/prolog/latest/Dockerfile
Comment (cjreed121) [inline dockerfiles/prolog/latest/Dockerfile:10]: Are these necessary? By installing `swi-prolog`, it should fetch any needed dependencies. But not sure what you plan on running in this image as it is possible you could need depending how you are using it.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: dockerfiles/prolog/latest/Dockerfile
Comment (ION606) [reply dockerfiles/prolog/latest/Dockerfile:10]: I noticed that sometimes prolog refused to run without them so I included them and added the --no-recommend flag

Comment (cjreed121) [reply dockerfiles/prolog/latest/Dockerfile:10]: Do you happen to have any reproducible examples of this? I'm okay with leaving them if they help but just wondering if there was clear examples of issues without it.

Comment (ION606) [reply dockerfiles/prolog/latest/Dockerfile:10]: huh. So I re-ran it and I was mistaken. Prolog works fine without them but the time to build the docker image goes from 22.9s to 44.9s with just
```Dockerfile
RUN apt-get update \
    && apt-get -y install swi-prolog \
    && rm -rf /var/lib/apt/lists/*
```

Comment (cjreed121) [reply dockerfiles/prolog/latest/Dockerfile:10]: Probably missing `--no-install-recommends`?

Comment (ION606) [reply dockerfiles/prolog/latest/Dockerfile:10]: would that not install needed things tho?

Comment (cjreed121) [reply dockerfiles/prolog/latest/Dockerfile:10]: That depends what you need. `--no-install-recommends` will still installed all main required dependencies but not optional ones. If you need the optional ones, then you'll have to leave it off or we just leave what code you currently have.

Comment (ION606) [reply dockerfiles/prolog/latest/Dockerfile]: no it seems to work fine, I wonder why I added it, maybe I changed something else that fixed it at the same time? In any case I updated the fork (and I also have to wake up in 5 hours so unfortunately I'm also going to bed)

**Comment by cjreed121** [comment]
Last thing is to change the `latest` dir name to `8` so the tag is `submitty/prolog:8`. Also want to wait until #44 is merged, then have you merge main back in so that the test runs properly.

**Review by cjreed121 — APPROVED** [review]

**Pull request merged**
